### PR TITLE
MINIFICPP-637 - extend C API to support adding processors with config…

### PR DESCRIPTION
**Pull request by arpadboda**
…uration to existing flows

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/capi/Plan.h [hunk 1]
```diff
@@ -79,6 +79,8 @@ class ExecutionPlan {
     return content_repo_;
   }
 
+  static std::shared_ptr<core::Processor> createProcessor(const std::string &processor_name, const std::string &name);
```
[the remark this answers is not in the document]
Comment (arpadboda) [reply libminifi/include/capi/Plan.h]: This is in Plan, where addProcessor also requires name to be specified, so the API is consistent in my opinion. 
I get what you mean, but in that case I would also prefer to modify Plan::addProcessor to have "name" as an optional argument.

File: libminifi/src/capi/api.cpp [hunk 2]
```diff
@@ -287,6 +288,41 @@ flow *create_getfile(nifi_instance *instance, flow *parent_flow, GetFileConfig *
   return new_flow;
 }
 
+/**
+ * Adds processor to existing flow
+ * @param proc_name: name of the new processor
+ * @param config: config of the processor, key=value pairs separated by ';'
+ * @return:
+ *   -0: success
+ *   -1: processor cannot be instanciated
```
[the remark this answers is not in the document]
Comment (arpadboda) [reply libminifi/src/capi/api.cpp]: Good shout, will change to signed.

File: libminifi/src/capi/api.cpp [hunk 3]
```diff
@@ -287,6 +288,41 @@ flow *create_getfile(nifi_instance *instance, flow *parent_flow, GetFileConfig *
...
+ *   -2: invalid config string
+ *   -3: invalid property name in config string
+ */
+uint8_t add_processor(flow *parent_flow, const char *proc_name, const char *config) {
```
[the remark this answers is not in the document]
Comment (arpadboda) [reply libminifi/src/capi/api.cpp]: Guess I was a bit confused by the current implementation of "create_getfile", which instantiates and sets the properties in one shot. I will change to do these separately.

[the remark this answers is not in the document]
Comment (phrocker) [reply libminifi/src/capi/api.cpp]: Sorry I see that more as a helper. If we're going down that route we should probably define more strongly typed configs and put that just outside of the core API

File: libminifi/src/capi/api.cpp [hunk 4]
```diff
@@ -287,6 +288,24 @@ flow *create_getfile(nifi_instance *instance, flow *parent_flow, GetFileConfig *
...
+    processor *new_processor = new processor();
+    new_processor->processor_ptr = proc.get();
+    return new_processor;
+  }
+  return nullptr;
+}
+int set_property(processor *proc, const char *name, const char *value) {
+  if (name != nullptr && value != nullptr && proc != nullptr) {
+    core::Processor *p = static_cast<core::Processor*>(proc->processor_ptr);
+    return p->setProperty(name, value) ? 0 : -1;
```
Comment (phrocker) [inline libminifi/src/capi/api.cpp]: can we return a different response code here? maybe -2? If we return anything here other than zero, probably should differentiate from invalid arguments.

**Pull request closed without merging**